content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,388 @@
+
+The `getDerivedStateFromProps` lifecycle was introduced in 16.3. At the time, we provided [some examples](/blog/2018/03/27/update-on-async-rendering.html#examples) of how to use the new lifecycle to derive state from props. Our primary goal was to help people migrate from legacy lifecycles to newer ones that are safer to use with the [upcoming async rendering mode](blog/2018/03/01/sneak-peek-beyond-react-16.html).
+
+Since that blog post was written, we've observed some common usage patterns for both the legacy `componentWillReceiveProps` and the newer `getDerivedStateFromProps` that cause bugs or otherwise confusing behavior. The 16.4 bugfix [makes this behavior more predictable](https://github.com/facebook/react/issues/12898), but the fixes aren't always obvious. Hopefully this blog post will help!
+
+Let's take a look at the following topics:
+* [When should I use derived state?](#when-should-i-use-derived-state)
+* [When should I avoid derived state?](#when-should-i-avoid-derived-state)
+* [What about memoization?](#what-about-memoization)
+
+## When should I use derived state?
+
+`getDerivedStateFromProps` exists for only one purpose. It enables a component to update its internal state as the result of **changes in external props**. This distinction is important. The problems we've seen so far can be ultimately reduced to either (1) unconditionally updating state from props or (2) updating state whenever props and state don't match. (We'll go over both in more detail below.)
+
+So what does it mean to update state as the result of changes in props? Our previous blog post provided a few examples, like [managing the current scroll direction based on an offset prop](/blog/2018/03/27/update-on-async-rendering.html#updating-state-based-on-props) or [loading external data specified by a source prop](/blog/2018/03/27/update-on-async-rendering.html#fetching-external-data-when-props-change).
+
+As a general rule, if you're not sure about whether you should use `getDerivedStateFromProps`, here are some questions to ask yourself:
+* Is the state **derived** from props (as opposed to just mirroring it)?
+* Is the state update specifically triggered by a **props change** (and not just the current props value)?
+
+If your answer to either of the above questions is "no" then there are better patterns to use.
+
+## When should I avoid derived state?
+
+The terms ["controlled"](/docs/forms.html#controlled-components) and ["uncontrolled"](/docs/uncontrolled-components.html) are often used to refer to form components, but they can also be used to describe where a component's state lives. Data that is passed in as props can be thought of as **controlled** and one that exists only in internal state can be thought of as **uncontrolled**.
```

Comment:
As I suggested in https://github.com/reactjs/reactjs.org/pull/931#discussion_r193184074, I think even a small elaboration on *why* one is called "controlled" and another is not would help. (Mine was probably too lengthy.) I don't think it's obvious props = controlled because parent "controls" them. Even saying this once would be helpful IMO.

Reply:
Oh, whoops. I think I missed that point somehow. I just noticed the "data" part. I'll squeeze a small explanation in.